File: CSharpAthletic/Network/ObjectProtocol/ClientWorker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using Domain;
using Domain.DTO;
using Services;

namespace Network.ObjectProtocol
{
    public class ClientWorker : IObserver
    {
        private readonly IServices Server;
        private readonly TcpClient Connection;
        private readonly NetworkStream Stream;
        private readonly IFormatter Formatter;
        private volatile bool Connected;

        public ClientWorker(IServices server, TcpClient connection)
        {
            Server = server;
            Connection = connection;
            try
            {
                Stream = connection.GetStream();
                Formatter = new BinaryFormatter();
                Connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public virtual void RefreshEvents(IEnumerable<EventCountDTO> eventCountList)
        {
            try
            {
                Console.WriteLine("Refresh Event Count List Response");
                EventCountDTO[] lst = eventCountList.ToArray();
                SendResponse(new RefreshEventsResponse { EventCountList = lst });
            }
            catch (Exception e)
            {
                throw new Exception(e.StackTrace);
            }
        }

        public virtual void Run()
        {
            while (Connected)
            {
                try
                {
                    object request = Formatter.Deserialize(Stream);
                    object response = HandleRequest((Request)request);
                    if (response != null)
                    {
                        SendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                Stream.Close();
                Connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        private Response HandleRequest(Request request)
        {
            if (request is LoginRequest)
            {
                Console.WriteLine("Login request");
                LoginRequest logReq = (LoginRequest)request;
                User user = logReq.User;
                try
                {
                    lock (Server)
                    {
                        Server.Login(user, this);
                    }
                    return new OkResponse();
                }
                catch (Exception e)
                {
                    Connected = false;
                    return new ErrorResponse { ErrorMessage = e.Message };
                }
            }

            if (request is LogoutRequest)
            {
                Console.WriteLine("Logout request");
                LogoutRequest logReq = (LogoutRequest)request;
                User user = logReq.User;
                try
                {
                    lock (Server)
                    {
                        Server.Logout(user, this);
                    }
                    Connected = false;
                    return new OkResponse();

                }
                catch (Exception e)
                {
                    return new ErrorResponse { ErrorMessage = e.Message };
                }
            }

            if (request is ChildrenEventsRequest)
            {
                Console.WriteLine("Children Events Request");
                ChildrenEventsRequest ceReq = (ChildrenEventsRequest)request;
                EventAgeGroupDTO eag = ceReq.EventAgeGroup;
                try
                {
                    ChildNoEventsDTO[] lst;
                    lock (Server)
                    {
                        lst = Server.GetChildrenNoEvents(eag.Event, eag.AgeGroup).ToArray();
                    }
                    return new ChildrenEventsResponse { ChildNoEventsList = lst };
                }
                catch (Exception e)
                {
                    return new ErrorResponse { ErrorMessage = e.Message };
                }
            }

            if (request is AddParticipantRequest)
            {
                Console.WriteLine("Add Participant Request");
                AddParticipantRequest apReq = (AddParticipantRequest)request;
                ParticipantDetailsDTO pd = apReq.ParticipantDetails;
                try
                {
                    lock (Server) 
                    {
                        Server.AddParticipant(pd.Name, pd.Age, pd.Event1, pd.Event2);
                    }
                    return new OkResponse();
                }
                catch (Exception e)
                {
                    return new ErrorResponse { ErrorMessage = e.Message };
                }
            }

            if (request is EventCountRequest)
            {
                Console.WriteLine("Event Count Request");
                try
                {
                    EventCountDTO[] lst;
                    lock (Server)
                    {
                        lst = Server.GetEventsNumber().ToArray();
                    }
                    return new EventCountResponse { EventCountList = lst };
                }
                catch (Exception e)
                {
                    return new ErrorResponse { ErrorMessage = e.Message };
                }
            }

            return null;
        }

        private void SendResponse(Response response)
        {
            lock (Stream)
            {
                Formatter.Serialize(Stream, response);
                Stream.Flush();
            }
        }
    }
}

File: JavaAthletic/AthleticNetwork/src/main/java/ClientWorker.java
import domain.User;
import dtos.EventAgeGroupDTO;
import dto.ParticipantDetailsDTO;
import dtos.ChildNoEventsDTO;
import dtos.EventCountDTO;
import protocol.Request;
import protocol.RequestType;
import protocol.Response;
import protocol.ResponseType;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.List;

public class ClientWorker implements Runnable, IObserver {
    private final IServiceServer server;
    private final Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ClientWorker(IServiceServer server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    @Override
    public void refreshEvents(List<EventCountDTO> eventCountList) throws Exception {
        EventCountDTO[] arr = eventCountList.toArray(new EventCountDTO[0]);
        Response response = new Response.Builder().type(ResponseType.REFRESH_EVENTS).data(arr).build();
        try {
            sendResponse(response);
        }
        catch (IOException e) {
            throw new Exception("Sending error\n" + e);
        }
    }

    @Override
    public void run() {
        while(connected) {
            try {
                Object request = input.readObject();
                Response response = handleRequest((Request)request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error\n" + e);
        }
    }

    private void sendResponse(Response response) throws IOException {
        System.out.println("Sending response " + response);
        output.writeObject(response);
        output.flush();
    }

    private final static Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    private Response handleRequest(Request request) {
        if (request.type() == RequestType.LOGIN) {
            System.out.println("Login request");
            User user = (User) request.data();
            try {
                server.login(user, this);
                return okResponse;
            } catch (Exception e) {
                connected = false;
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        if (request.type() == RequestType.LOGOUT) {
            System.out.println("Logout request");
            User user = (User) request.data();
            try {
                server.logout(user, this);
                connected = false;
                return okResponse;
            } catch (Exception e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        if (request.type() == RequestType.CHILDREN_EVENTS) {
            System.out.println("Children events request");
            EventAgeGroupDTO eag = (EventAgeGroupDTO) request.data();
            try {
                List<ChildNoEventsDTO> lst = server.getChildrenNoEvents(eag.getEvent(), eag.getAgeGroup());
                ChildNoEventsDTO[] arr = lst.toArray(new ChildNoEventsDTO[0]);
                return new Response.Builder().type(ResponseType.CHILDREN_EVENTS).data(arr).build();
            }
            catch (Exception e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        if (request.type() == RequestType.ADD_PARTICIPANT) {
            System.out.println("Add participant request");
            ParticipantDetailsDTO pd = (ParticipantDetailsDTO) request.data();
            try {
                server.addParticipant(pd.getName(), pd.getAge(), pd.getEvent1(), pd.getEvent2());
                return okResponse;
            } catch (Exception e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        if (request.type() == RequestType.EVENT_COUNT) {
            System.out.println("Event count request");
            try {
                List<EventCountDTO> lst = server.getEventsNumber();
                EventCountDTO[] arr = lst.toArray(new EventCountDTO[0]);
                return new Response.Builder().type(ResponseType.EVENT_COUNT).data(arr).build();

            } catch (Exception e) {
                return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
            }
        }
        return null;
    }

}
